Changed weights initialization & added stride functionality #100

Ujjwal-9 opened the pull request:
Initialization of all weights to zero mathematically leads to either the neuron values are zero (for multi layers) or the delta would be zero.

Ujjwal-9 commented:
@iamtrask pls look into this PR.

AkashGutha commented on syft/tensor.py:
```diff
@@ -1372,6 +1374,30 @@ def mm(self, tensor2):
         return syft.mm(self, tensor2)
 
+    def stride(self, dim=None):
+        """Returns the jump necessary to go from one element to the next one in the specified dimension dim."""
```
Please make the documentation consistent with #252

you can take a look at this link 
https://github.com/OpenMined/PySyft/blob/4e2aa9e34df985f4d8e0703d900eaa2558fd413a/syft/math.py#L259

Ujjwal-9 commented:
The normal vs uniform init seem to be rather unclear. But i have changed my initialization to uniform. 
@trask look into it.

AkashGutha commented on syft/tensor.py:
```diff
@@ -1372,6 +1374,30 @@ def mm(self, tensor2):
+            out = calculation(ndim, shp)
+            return out[dim]
+        else:
+            raise RuntimeError("dimension out of range of {}D tensor".format(ndim))
```
Runtime error!! I guess not.
We can do a simple `Error` or better `ValueError` which suits this situation.
Look into these. `RuntimeError` is definitely not the one we should be using.

Ujjwal-9 replied:
Pytorch used Runtime error for this issue.

AkashGutha commented:
This PR fixes #100 
adding #100 in the title doesn't actually create reference adding it here will do.

AkashGutha commented on syft/tensor.py:
```diff
@@ -1397,6 +1399,35 @@ def mm(self, tensor2):
+        """
+        Returns the jump necessary to go from one element to the next one in the specified dimension dim.
+
+        :param dim: dimension
```
This still hasn't resonated with the expected style of documentation.
use Parameters and Results and seperators
https://github.com/OpenMined/PySyft/blob/4e2aa9e34df985f4d8e0703d900eaa2558fd413a/syft/math.py#L259

aiorla commented:
Thank you @Ujjwal-9! :+1: 

First of all I'd like to say that I prefer different PR's for unrelated changes to improve its reviewing.
* Weights initialization
    *  I'm with you that having zero-initialization as the default can be sometimes a bit worrying.
    * Not sure about this normal vs uniform. Commonly, e.g. [PyTorch](http://pytorch.org/docs/master/_modules/torch/nn/modules/linear.html#Linear) or [keras](https://keras.io/initializers/), the weights are initialized with a normal/uniform with a std/limit dependent on the number of features. I'd recommend following again PyTorch example here (although it's not that important).
* `stride` functionality
    * I think that your function does not provide the same functionality than PyTorch (where only a tuple is returned and no parameter "dim" is needed). If your version is a function that does what PyTorch does + extra functionalities with the optional parameter `dim` please tell me and I'll look it more carefully.

Ujjwal-9 commented:
@aiorla if you look at definition of stride in pytorch source code you will find that they provide `dim` parameter. Actually dim is not necessary parameter. checkout stride() definition [here](https://github.com/pytorch/pytorch/blob/312e0ce3ba1fa74854a1fc82aa1b7c24baeb4006/torch/lib/TH/generic/THTensor.c#L28)
You can also try this code:
```
import torch
a = torch.Tensor([[1,2,3,4,5],[6,7,8,9,10]])
a.stride()
a.stride(0)
a.stride(1)
a.stride(-1)
a.stride(-2)
```

aiorla commented:
It looks like PyTorch [documentation](http://pytorch.org/docs/master/tensors.html#torch.Tensor.stride) is incomplete then, sorry @Ujjwal-9... :sweat_smile:
I'll give it another look later then (if nobody does it before me). :+1:

Ujjwal-9 commented:
@aiorla I checked Numpy implementation but it gives different results. 
And sorry i didn't write unit test. will write it asap. :sweat_smile:

aiorla commented:
@Ujjwal-9 numpy implementation gives the strides in bytes, so a division by the element size is needed. Does it have more differences? (I haven't thoroughly checked)

sp4ke commented:
@Ujjwal-9 @aiorla   I am currently working on the storage features from #93 and it requires the `strides` features following PyTorch signature. Since it is already implemented here, should I rebase over this PR ?

sp4ke commented:
Forgot one question on the `strides` implementation, why not reuse the underlying strides from the ndarray in `self.data` ?

aiorla commented:
Hi @sp4ke! I talked last week with @Ujjwal-9 on Slack and he told me that he was going to use `strides()` & add the unit test but he might have been busy.

I'm not sure what you should do there (if you don't want to wait for the fix). If I were you I'd maybe implement an inlined version of strides inside `set_()` and add a TODO to use `strides()` when available... (but that may not be the "cleanest" way :sweat_smile:)

sp4ke commented:
Ha ! I will do that at least so I can progress. I think by the time my PR would be approved strides will be available. 

By the way do you know any document that describes the "architecture" and API of the pysyft library in regard to PyTorch and Numpy. Right now it seems a bit messy, some methods reuse the internal ndarray features and some others re-implement them

bharathgs commented:
@sp4ke as of now, I think there's no architecture as such. however, there are a few of these inconsistencies which are being discussed in issue #256 which needs to be addressed.

bharathgs commented:
@Ujjwal-9 any updates on this?

Ujjwal-9 commented:
@bharathgs its complete now you can review it.
changes:
1.  Changed my custom stride function to numpy implementation
2.  Added Tests
3.  Changed weights initialization in linear.py

bharathgs commented on syft/tensor.py:
```diff
@@ -3331,6 +3331,31 @@ def renorm_(self, p, dim, maxnorm):
+
+        Returns
+        -------
+        Tuple
```
A tuple is returned when there's no Argument, If not an int is returned.
so maybe you can change the doc to reflect that?

Ujjwal-9 commented:
Made the changes
